lib/package_json.rb

```diff
@@ -79,16 +79,23 @@ def determine_package_manager(fallback_manager)
 
     name, version = package_manager.split("@")
 
-    if name == "yarn"
-      raise Error, "a major version must be present for Yarn" if version.nil? || version.empty?
-      return :yarn_classic if version.start_with?("1")
-
-      return :yarn_berry
-    end
+    return determin_yarn_version(version) if name == "yarn"
 
     name.to_sym
   end
 
+  def determin_yarn_version(version)
+    raise Error, "a major version must be present for Yarn" if version.nil? || version.empty?
+
+    # Regex for matching version 1 with ~ or ^ prefix and optionally minor and path version
+    # This regext doesn't check the validity of the version format!
+    version_1_regex = /^[~^]?1(\.|$)/
+
+    return :yarn_classic if version.match? version_1_regex
```

Comment:
This is looking good though I still don't think we need the dedicated variable, and the comment should be updated:

```suggestion
    # check to see if we're meant to be using Yarn v1 based on the versions major component,
    # and accounting for the presents of version constraints like ^, ~, and =
    return :yarn_classic if version.match? /^[^~=]?1(\.|$)/
```

(I've also updated the regexp to include `=` and put the constraints in order of their specificity from less to exact a little bonus)